File: ql/time/calendars/unitedkingdom.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*
 Copyright (C) 2004 Ferdinando Ametrano
 Copyright (C) 2000, 2001, 2002, 2003 RiskMap srl

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
*/

#include <ql/time/calendars/unitedkingdom.hpp>
#include <ql/errors.hpp>

namespace QuantLib {

    namespace {

        // common rules

        bool isBankHoliday(Day d, Weekday w, Month m, Year y) {
            return
                // first Monday of May (Early May Bank Holiday)
                // moved to May 8th in 1995 and 2020 for V.E. day
                (d <= 7 && w == Monday && m == May && y != 1995 && y != 2020)
                || (d == 8 && m == May && (y == 1995 || y == 2020))
                // last Monday of May (Spring Bank Holiday)
                // moved to in 2002, 2012 and 2022 for the Golden, Diamond and Platinum
                // Jubilee with an additional holiday
                || (d >= 25 && w == Monday && m == May && y != 2002 && y != 2012 && y != 2022)
                || ((d == 3 || d == 4) && m == June && y == 2002)
                || ((d == 4 || d == 5) && m == June && y == 2012)
                || ((d == 2 || d == 3) && m == June && y == 2022)
                // last Monday of August (Summer Bank Holiday)
                || (d >= 25 && w == Monday && m == August)
                // April 29th, 2011 only (Royal Wedding Bank Holiday)
                || (d == 29 && m == April && y == 2011)
                // September 19th, 2022 only (The Queen's Funeral Bank Holiday)
                || (d == 19 && m == September && y == 2022)
                ;
        }

    }

    UnitedKingdom::UnitedKingdom(UnitedKingdom::Market market) {
        // all calendar instances on the same market share the same
        // implementation instance
        static ext::shared_ptr<Calendar::Impl> settlementImpl(
                                           new UnitedKingdom::SettlementImpl);
        static ext::shared_ptr<Calendar::Impl> exchangeImpl(
                                           new UnitedKingdom::ExchangeImpl);
        static ext::shared_ptr<Calendar::Impl> metalsImpl(
                                           new UnitedKingdom::MetalsImpl);
        switch (market) {
          case Settlement:
            impl_ = settlementImpl;
            break;
          case Exchange:
            impl_ = exchangeImpl;
            break;
          case Metals:
            impl_ = metalsImpl;
            break;
          default:
            QL_FAIL("unknown market");
        }
    }

    bool UnitedKingdom::SettlementImpl::isBusinessDay(const Date& date) const {
        Weekday w = date.weekday();
        Day d = date.dayOfMonth(), dd = date.dayOfYear();
        Month m = date.month();
        Year y = date.year();
        Day em = easterMonday(y);
        if (isWeekend(w)
            // New Year's Day (possibly moved to Monday)
            || ((d == 1 || ((d == 2 || d == 3) && w == Monday)) &&
                m == January)
            // Good Friday
            || (dd == em-3)
            // Easter Monday
            || (dd == em)
            || isBankHoliday(d, w, m, y)
            // Christmas (possibly moved to Monday or Tuesday)
            || ((d == 25 || (d == 27 && (w == Monday || w == Tuesday)))
                && m == December)
            // Boxing Day (possibly moved to Monday or Tuesday)
            || ((d == 26 || (d == 28 && (w == Monday || w == Tuesday)))
                && m == December)
            // December 31st, 1999 only
            || (d == 31 && m == December && y == 1999))
            return false; // NOLINT(readability-simplify-boolean-expr)
        return true;
    }


    bool UnitedKingdom::ExchangeImpl::isBusinessDay(const Date& date) const {
        Weekday w = date.weekday();
        Day d = date.dayOfMonth(), dd = date.dayOfYear();
        Month m = date.month();
        Year y = date.year();
        Day em = easterMonday(y);
        if (isWeekend(w)
            // New Year's Day (possibly moved to Monday)
            || ((d == 1 || ((d == 2 || d == 3) && w == Monday)) &&
                m == January)
            // Good Friday
            || (dd == em-3)
            // Easter Monday
            || (dd == em)
            || isBankHoliday(d, w, m, y)
            // Christmas (possibly moved to Monday or Tuesday)
            || ((d == 25 || (d == 27 && (w == Monday || w == Tuesday)))
                && m == December)
            // Boxing Day (possibly moved to Monday or Tuesday)
            || ((d == 26 || (d == 28 && (w == Monday || w == Tuesday)))
                && m == December)
            // December 31st, 1999 only
            || (d == 31 && m == December && y == 1999))
            return false; // NOLINT(readability-simplify-boolean-expr)
        return true;
    }


    bool UnitedKingdom::MetalsImpl::isBusinessDay(const Date& date) const {
        Weekday w = date.weekday();
        Day d = date.dayOfMonth(), dd = date.dayOfYear();
        Month m = date.month();
        Year y = date.year();
        Day em = easterMonday(y);
        if (isWeekend(w)
            // New Year's Day (possibly moved to Monday)
            || ((d == 1 || ((d == 2 || d == 3) && w == Monday)) &&
                m == January)
            // Good Friday
            || (dd == em-3)
            // Easter Monday
            || (dd == em)
            || isBankHoliday(d, w, m, y)
            // Christmas (possibly moved to Monday or Tuesday)
            || ((d == 25 || (d == 27 && (w == Monday || w == Tuesday)))
                && m == December)
            // Boxing Day (possibly moved to Monday or Tuesday)
            || ((d == 26 || (d == 28 && (w == Monday || w == Tuesday)))
                && m == December)
            // December 31st, 1999 only
            || (d == 31 && m == December && y == 1999))
            return false; // NOLINT(readability-simplify-boolean-expr)
        return true;
    }

}

File: quantlib/time/calendars/unitedkingdom.py
from __future__ import annotations

from enum import IntEnum

from quantlib.time.calendar import Calendar
from quantlib.time.date import Date
from quantlib.time.month import Month
from quantlib.time.weekday import Weekday


class UnitedKingdom(Calendar):
    class Market(IntEnum):
        Settlement = 0
        Exchange = 1
        Metals = 2

    class _SettlementImpl(Calendar.WesternImpl):
        def name(self) -> str:
            return "UK settlement"

        def isBusinessDay(self, date: Date) -> bool:
            w = date.weekday()
            d = date.dayOfMonth()
            dd = date.dayOfYear()
            m = date.month()
            y = date.year()
            em = self.easterMonday(y)

            if (
                self.isWeekend(w)
                # New Year's Day (possibly moved to Monday)
                or ((d == 1 or ((d == 2 or d == 3) and w == Weekday.Monday)) and m == Month.January)
                # Good Friday
                or (dd == em - 3)
                # Easter Monday
                or (dd == em)
                # first Monday of May (Early May Bank Holiday)
                or (d <= 7 and w == Weekday.Monday and m == Month.May)
                # last Monday of May (Spring Bank Holiday)
                or (d >= 25 and w == Weekday.Monday and m == Month.May and y != 2002 and y != 2012 and y != 2022)
                # Spring Bank Holiday special cases
                or ((d == 3 or d == 4) and m == Month.June and y == 2002)
                or ((d == 4 or d == 5) and m == Month.June and y == 2012)
                or ((d == 2 or d == 3) and m == Month.June and y == 2022)
                # Summer Bank Holiday (last Monday of August)
                or (d >= 25 and w == Weekday.Monday and m == Month.August)
                # September 19, 2022 - Queen Elizabeth II funeral
                or (d == 19 and m == Month.September and y == 2022)
                # Christmas (possibly moved to Monday or Tuesday)
                or ((d == 25 or (d == 27 and (w == Weekday.Monday or w == Weekday.Tuesday))) and m == Month.December)
                # Boxing Day (possibly moved to Monday or Tuesday)
                or ((d == 26 or (d == 28 and (w == Weekday.Monday or w == Weekday.Tuesday))) and m == Month.December)
                # May 8, 2020 - 75th anniversary of VE day
                or (d == 8 and m == Month.May and y == 2020)
                # May 8, 2023 - Coronation of King Charles III
                or (d == 8 and m == Month.May and y == 2023)
            ):
                return False
            return True

    def __init__(self, market: UnitedKingdom.Market = Market.Settlement):
        if market == UnitedKingdom.Market.Settlement:
            impl: Calendar.Impl = UnitedKingdom._SettlementImpl()
        else:
            raise ValueError(f"UK calendar market {market} not yet implemented")
        super().__init__(impl)
